testplan/runnable/interactive/http.py

```diff
@@ -1,21 +1,284 @@
 """
 Http handler for interactive mode.
 """
+from __future__ import unicode_literals
+from __future__ import print_function
+from __future__ import division
+from __future__ import absolute_import
 
-import json
-import six
+from builtins import super
+from builtins import str
+from future import standard_library
+
+standard_library.install_aliases()
 import uuid
-import inspect
-import threading
+from multiprocessing import dummy
 
-from six.moves.BaseHTTPServer import HTTPServer, BaseHTTPRequestHandler
-from six.moves.socketserver import ThreadingMixIn
-from six.moves.urllib.parse import urlparse
+import flask
+import flask_restplus
+from cheroot import wsgi
+import werkzeug.exceptions
+import marshmallow.exceptions
 
-from testplan.common.utils.exceptions import format_trace
 from testplan.common.config import ConfigOption
 from testplan.common.entity import Entity, EntityConfig, RunnableIHandler
 from testplan import defaults
+from testplan import report
+
+
+def generate_interactive_api(http_handler, ihandler):
+    """Generates the interactive API using Flask."""
+    app = flask.Flask(__name__)
+    api = flask_restplus.Api(app)
+
+    @api.route("/api/v1/interactive/report")
+    class Report(flask_restplus.Resource):
+        """
+        Interactive report endpoint. There is a single root report object
+        for interactive mode.
+        """
+
+        def get(self):
+            """Get the state of the root interactive report."""
+            with ihandler.report_mutex:
+                return ihandler.report.shallow_serialize()
+
+        def put(self):
+            """Update the state of the root interactive report."""
+            if flask.request.json is None:
+                raise werkzeug.exceptions.BadRequest(
+                    "JSON body is required for PUT"
+                )
+
+            with ihandler.report_mutex:
+                try:
+                    new_report = report.TestReport.shallow_deserialize(
+                        flask.request.json, ihandler.report
+                    )
+                except marshmallow.exceptions.ValidationError as e:
+                    raise werkzeug.exceptions.BadRequest(str(e))
+
+                if should_run(ihandler.report.status):
+                    new_report.status = report.Status.RUNNING
+                    http_handler.pool.apply_async(ihandler.run_tests)
+
+                ihandler.report = new_report
+                return ihandler.report.shallow_serialize()
+
+    @api.route("/api/v1/interactive/report/tests")
+    class Tests(flask_restplus.Resource):
+        """
+        Tests endpoint. Represents all Test objects in the report. Read-only.
+        """
+
+        def get(self):
+            """Get the UIDs of all tests defined in the testplan."""
+            with ihandler.report_mutex:
+                return [test.uid for test in ihandler.report]
```

Comment:
The lock is required when reading/updating the report object, since otherwise other threads could mutate the report object halfway through when we are iterating through the tests it contains.